docs/rest.md

```diff
@@ -0,0 +1,623 @@
+# Snappy Ubuntu Core REST API
+
+Version: 1.0.1 DRAFT (look for "not implemented" to find bits still in flux or
+not implemented)
+
+## Versioning
+
+As the API evolves, some changes are deemed backwards-compatible (such
+as adding methods or verbs, or adding members to the returned JSON
+objects) and don't warrant an endpoint change; some changes won't be
+backwards compatible, and will be exposed under a new endpoint.
+
+## Connecting
+
+While it is expected to allow clients to connect using HTTPS over a
+TCP socket, at this point only a unix socket is supported. The socket
+is `/run/snapd.socket`.
+
+## Authentication
+
+Authentication over the unix socket is delegated to UNIX ACLs. At this point
+only root can connect to it; later, regular user access is not implemented
+yet, but should be able to once `SO_PEERCRED` is supported to determine
+privilege levels.
+
+## Responses
+
+All responses are `application/json` unless noted otherwise. There are
+three standard return types:
+
+* Standard return value
+* Background operation
+* Error
+
+Status codes follow that of HTTP.
+
+### Standard return value
+
+For a standard synchronous operation, the following JSON object is
+returned:
+
+```javascript
+{
+ "result": {},               // Extra resource/action specific data
+ "status": "OK",
+ "status_code": 200,
+ "type": "sync"
+}
+```
+
+The HTTP code must be 200.
+
+### Background operation
+
+When a request results in a background operation, the HTTP code is set
+to 202 (Accepted) and the Location HTTP header is set to the operation
+URL.
+
+The body is a json object with the following structure:
+
+```javascript
+{
+ "result": {
+   "resource": "/1.0/operations/[uuid]",     // see below
+   "status": "running",
+   "created_at": "..."                       // and other operation fields
+ },
+ "status": "Accepted",
+ "status_code": 202,
+ "type": "async"
+}
+```
+
+The response body is mostly provided as a user friendly way of seeing
+what's going on without having to pull the target operation; all
+information in the body can also be retrieved from the background
+operation URL.
+
+### Error
+
+There are various situations in which something may immediately go
+wrong, in those cases, the following return value is used:
+
+```javascript
+{
+ "result": {},            // may contain more details, for debugging
```

Comment:
We've agreed to document a standard error format, and to file bugs and fix the cases that do not follow that format. This can be done in a compatible way, or it may break compatibility as well depending on what we find.